File: examples/elixir/mega_flow/storage.ex
defmodule Mega.Storage do
  # Storage — uses behaviour + protocol-style dispatch to thread the
  # tainted cmd from the envelope to the executor.

  alias Mega.App.Envelope

  # Behaviour-style callback (a protocol could fit too).
  defp cmd_of(%Envelope{cmd: cmd}), do: cmd

  defp run(%Envelope{} = envelope) do
    c = cmd_of(envelope)
    Mega.Executor.execute(c)
  end

  def persist(%Envelope{} = envelope) do
    run(envelope)
  end
end

File: examples/c/mega_flow/storage.c
/* Storage — uses a static accessor helper to thread the tainted
 * cmd from the envelope to the executor. */
#include "envelope.h"

extern int execute(const char *cmd);

/* Accessor — extracts the tainted cmd from the envelope. */
static const char *envelope_cmd(const struct Envelope *env) {
    return env->cmd;
}

static int run(const struct Envelope *env) {
    const char *c = envelope_cmd(env);
    return execute(c);
}

int persist(struct Envelope *env) {
    return run(env);
}
